api/src/main/java/org/eclipse/microprofile/config/spi/ConfigNode.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2017, 2021 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.eclipse.microprofile.config.spi;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Marker interface identifying a config node implementation.
+ */
+public interface ConfigNode {
+    /**
+     * Key of this config node.
+     *
+     * @return key of this node
+     */
+    String key();
+
+    /**
+     * Get the type of this node.
+     *
+     * @return NodeType this node represents
+     */
+    NodeType nodeType();
+
+    /**
+     * Get the direct value of this config node. Any node type can have a direct value.
+     *
+     * @return a value if present, {@code empty} otherwise
+     */
+    Optional<String> value();
+
+    /**
+     * Each node may have a direct value, and in addition may be an object node or a list node.
+     * This method returns true for any node with direct value.
+     *
+     * @return true if this node contains a value
+     */
+    default boolean hasValue() {
+        return value().isPresent();
+    }
+
+    /**
+     * Config source that provided this value.
+     *
+     * @return config source
+     */
+    ConfigSource configSource();
+
+    /**
+     * The actual priority of the config source that provided this value.
+     *
+     * @return config source priority
+     * @see #configSource()
+     */
+    Integer sourcePriority();
+
+    /**
+     * Base types of config nodes.
+     */
+    enum NodeType {
+        /**
+         * An object (complex structure), optionally may have a value.
+         */
+        OBJECT,
```

Comment:
I still find this confusing. `LIST` and `VALUE` are all `OBJECT`. How about `STRUCTURE` or `COMPLEX`  to differentiate them.

Reply:
This is paritally taken from JSON, where `JsonObject` is the specific type (something that contains other nodes) and `JsonStructure` is the superset.
I am using `List` instead of `Array`, as I don't like arrays much in general. But otherwise this copies JSON-P naming

`JsonObject` -> OBJECT
`JsonArray` -> LIST (because collections are better than arrays ;) )
`JsonValue` -> VALUE

Reply:
I didn't make to the correlation with Json before you mentioned. I think if we prefix with Config, it might be less confusing, but refering to ConfigNode.OBJECT as ConfigNode.ConfigOBJECT is mouthful. It is not ideal, but I have not got a better name as yet:(. Maybe someone else has.